src/event.jl

```diff
@@ -94,23 +101,49 @@ Base.getindex(evt::CLEvent, evt_info::Symbol) = info(evt, evt_info)
     end
 end
 
-function event_notify(evt_id::CL_event, status::CL_int, julia_func::Ptr{Void})
-    # Obtain the Function object from the opaque pointer
-    callback = unsafe_pointer_to_objref(julia_func)::Function
+function event_notify(evt_id::CL_event, status::CL_int, payload::Ptr{Void})
+    ptrs = convert(Ptr{Ptr{Void}}, payload)
+    handle = unsafe_load(ptrs, 1)
+    ptr_evt_id = convert(Ptr{CL_event}, unsafe_load(ptrs, 2))
+    ptr_status = convert(Ptr{CL_int}, unsafe_load(ptrs, 3))
 
-    # In order to callback into the Julia thread create an AsyncWork package.
-    cb_packaged = Base.SingleAsyncWork(data -> callback(evt_id, status))
+    unsafe_store!(ptr_evt_id, evt_id, 1)
+    unsafe_store!(ptr_status, status, 1)
 
     # Use uv_async_send to notify the main thread
-    ccall(:uv_async_send, Void, (Ptr{Void},), cb_packaged.handle)
+    ccall(:uv_async_send, Void, (Ptr{Void},), handle)
+    nothing
 end
 
-const event_notify_ptr = cfunction(event_notify, Void,
-                                   (CL_event, CL_int, Ptr{Void}))
+function preserve_callback(evt :: CLEvent, cb, ptr)
+    evt._cbs[cb] = 0
+    push!(evt._memory, ptr)
+end
 
 
 function add_callback(evt::CLEvent, callback::Function)
-    @check api.clSetEventCallback(evt.id, CL_COMPLETE, event_notify_ptr, callback)
+    event_notify_ptr = cfunction(event_notify, Void,
+                                   (CL_event, CL_int, Ptr{Void}))
+
+    r_evt_id = Ref{CL_event}(0)
+    p_evt_id = Base.unsafe_convert(Ptr{CL_event}, r_evt_id)
+    r_status = Ref{CL_int}(0)
+    p_status = Base.unsafe_convert(Ptr{CL_int}, r_status)
+
+    cb = Base.SingleAsyncWork(data -> begin
+		println("Received callback")
```

Comment:
println isn't really allowed here. the SingleAsyncWork callback is being run directly from libuv's callback, so anything that may try to block (such as the flush at the end of printing) could result in a recursive call to the libuv event loop. a better pattern is for the SingleAsyncWork callback to notify a Condition variable, and do the real work on a thread that monitors that variable (a PR to Base to replace the callback with a Condition variable would be a useful improvement to this API).